core/src/Kokkos_Graph.hpp

```diff
@@ -0,0 +1,201 @@
+  std::shared_ptr<impl_t> m_impl_ptr = nullptr;
+
+  // </editor-fold> end private data members }}}2
+  //----------------------------------------------------------------------------
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="private ctors"> {{{2
+
+  // Note: only create_graph() uses this constructor, but we can't just make
+  // that a friend instead of GraphAccess because of the way that friend
+  // function template injection works.
+  explicit Graph(std::shared_ptr<impl_t> arg_impl_ptr)
+      : m_impl_ptr(std::move(arg_impl_ptr)) {}
+
+  // </editor-fold> end private ctors }}}2
+  //----------------------------------------------------------------------------
+
+ public:
+  ExecutionSpace const& get_execution_space() const {
+    return m_impl_ptr->get_execution_space();
+  }
+
+  void submit() const& {
+    KOKKOS_EXPECTS(bool(m_impl_ptr))
+    (*m_impl_ptr).submit();
+  }
+
+  void submit() && {
+    KOKKOS_EXPECTS(bool(m_impl_ptr))
+    // The graph interface isn't thread-safe, so we can rely on this
+    if (m_impl_ptr.use_count() == 1) {
+      std::move(*m_impl_ptr).submit();
+    } else {
+      (*m_impl_ptr).submit();
+    }
+  }
+};
+
+// </editor-fold> end Graph }}}1
+//==============================================================================
+
+//==============================================================================
+// <editor-fold desc="when_all"> {{{1
+
+template <class... PredecessorRefs>
+// constraints (not intended for subsumption, though...)
+//   ((remove_cvref_t<PredecessorRefs> is a specialization of
+//        GraphNodeRef with get_root().get_graph_impl() as its GraphImpl)
+//      && ...)
+auto when_all(PredecessorRefs&&... arg_pred_refs) {
+  // TODO @graph @desul-integration check the constraints and preconditions
+  //                                once we have folded conjunctions from
+  //                                desul
+  static_assert(sizeof...(PredecessorRefs) > 0,
+                "when_all() needs at least one predecessor.");
+  auto graph_ptr_impl =
+      Kokkos::Impl::GraphAccess::get_graph_weak_ptr(
+          std::get<0>(std::forward_as_tuple(arg_pred_refs...)))
+          .lock();
+  auto node_ptr_impl = graph_ptr_impl->create_aggregate_ptr(arg_pred_refs...);
+  graph_ptr_impl->add_node(node_ptr_impl);
+  KOKKOS_IMPL_FOLD_COMMA_OPERATOR(
+      graph_ptr_impl->add_predecessor(node_ptr_impl, arg_pred_refs) /* ... */);
+  return Kokkos::Impl::GraphAccess::make_graph_node_ref(
+      std::move(graph_ptr_impl), std::move(node_ptr_impl));
+}
+
+// </editor-fold> end when_all }}}1
+//==============================================================================
+
+//==============================================================================
+// <editor-fold desc="create_graph"> {{{1
+
+template <class ExecutionSpace, class Closure>
+Graph<ExecutionSpace> create_graph(ExecutionSpace ex, Closure&& arg_closure) {
```

Comment:
Yes. `ExecutionSpace` is already a reference type, and as a general rule I try to avoid taking reference types by reference unless I have a good reason to do so—besides being inefficient at times, it's confusing to the reader. At this granularity, there's no reason at all to worry about the cost of copying a pointer, so I went with the more readable option.